### DISCUSSION TITLE
[FEAT] Access ledger state while applying transactions to it rather than blocks?

### DISCUSSION THREAD

**Discussion by eyeinsky** [IntersectMBO/ouroboros-consensus#343] [Q&A] [answered]
Hi! This is very much a code-API question and just maybe a feature request :)

When writing a chain indexer that needs a full ledger state anyhow, I'd like to use the utxo set from within ledger state to resolve transaction inputs to their values. Thus far I've kept an (essentially duplicate) utxo set around. But the right thing to do would just use the one from ledger state.

The issue I'm facing is that [`tickThenReapplyLedgerResult`](https://github.com/input-output-hk/ouroboros-consensus/blob/f318ffeab455cef8f27c5cd9e4b898d895ed25df/ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Ledger/Abstract.hs#L146) applies blocks, so when there are transactions that do "intra-block spends" then these won't be seen neither in the initial nor the resulting ledger state.

Is it theoretically possible to access ledger state in between applying specific transactions, and also how ready is the current code base to achieve that (i.e maybe I just haven't found the right module yet, but it's already available somewhere)?

**Comment by amesgen** [comment, marked as answer]
Drive-by comment: maybe you are looking for [`applyTx`](https://input-output-hk.github.io/ouroboros-consensus/haddocks/ouroboros-consensus/Ouroboros-Consensus-Ledger-SupportsMempool.html#v:applyTx)?

**Comment by eyeinsky** [comment]
Thanks, I think that is the thing I was looking for! :)

**Comment by dnadales** [comment]
Thank you for your question Markus. I'll take the liberty to convert this issue into a discussion, so that we can improve our knowledge base. Let me know if you have any objections :pray:
